### PR TITLE
[MAINTENANCE] Enable "rule_based_profiler.run()" Method to Accept Batch Data Arguments Directly

### PR DESCRIPTION
**Pull request by alexsherstinsky**
### Scope
The logic of running profiler with data has to change in a way that reconciles the configuration first and then applies `Batch` data (as either list of `Batch` objects or `batch_request`, but not both) second.  This is important, because configuration has to be serializable, while `Batch` objects and/or `batch_request` are generally not.  However, the Rule-Based Profiler architecture mandates that all `Batch` objects and/or full `batch_request` is passed to all relevant `Builder` components (each of which contains internal logic for selecting `Batch` objects of interest).  This provides flexibility in running profiler.

### Notes
* Existing unit tests have been updated, changing the `Mock` points where necessary.
* Existing `profiler.add_rule()` tests have been updated, requiring the dictionary version of `batch_request` to be passed.

Please annotate your PR title to describe what the PR does, then give a brief bulleted description of your PR below. PR titles should begin with [BUGFIX], [FEATURE],  [DOCS], or [MAINTENANCE]. If a new feature introduces breaking changes for the Great Expectations API or configuration files, please also add [BREAKING]. You can read about the tags in our [contributor checklist](https://docs.greatexpectations.io/en/latest/contributing/contribution_checklist.html).

Changes proposed in this pull request:
- JIRA: GREAT-464/GREAT-498/GREAT-666
-
-

After submitting your PR, CI checks will run and @ge-cla-bot will check for your CLA signature.

For a PR with nontrivial changes, we review with both design-centric and code-centric lenses.

In a design review, we aim to ensure that the PR is consistent with our relationship to the open source community, with our software architecture and abstractions, and with our users' needs and expectations. That review often starts well before a PR, for example in github issues or slack, so please link to relevant conversations in notes below to help reviewers understand and approve your PR more quickly (e.g. `closes #123`).

Previous Design Review notes:
-
-
-

### Definition of Done
Please delete options that are not relevant.

- [x] My code follows the Great Expectations [style guide](https://docs.greatexpectations.io/docs/contributing/style_guides/code_style)
- [x] I have performed a [self-review](https://docs.greatexpectations.io/docs/contributing/contributing_checklist) of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added [unit tests](https://docs.greatexpectations.io/docs/contributing/contributing_test#writing-unit-and-integration-tests) where applicable and made sure that new and existing tests are passing.
- [x] I have run any local integration tests and made sure that nothing is broken.

Thank you for submitting!

### CODE REVIEW

File: great_expectations/validator/validator.py [hunk 1]
```diff
@@ -550,41 +546,34 @@ def _build_rule_based_profiler(
...
-        elif domain_type == MetricDomainTypes.TABLE:
-            pass  # No action is needed.
-        else:
-            pass  # No action is needed.
+        rule = profiler.generate_rule_overrides_from_batch_request(
+            rules=[rule],
+            batch_list=list(self.batches.values()),
+            batch_request=None,
+            force_batch_data=False,
+        )[0]
```
Comment (cdkini) [inline great_expectations/validator/validator.py:576]: Nitpick - could we split out the creation of the list and the retrieval of the first element into separate lines?

File: great_expectations/validator/validator.py [hunk 2]
```diff
@@ -595,15 +584,18 @@ def _validate_rule_and_update_rule_properties(
...
 
+        domain_type: MetricDomainTypes = rule.domain_builder.domain_type
+        # TODO: <Alex>Handle future domain_type cases as they are defined.</Alex>
+        if domain_type == MetricDomainTypes.COLUMN:
+            column_name = expectation_kwargs["column"]
+            rule.domain_builder.column_names = [column_name]
+        elif domain_type == MetricDomainTypes.TABLE:
+            pass  # No action is needed.
+        else:
+            pass  # No action is needed.
```
Comment (cdkini) [inline great_expectations/validator/validator.py]: logging?

File: great_expectations/validator/validator.py [hunk 3]
```diff
@@ -1425,7 +1417,7 @@ def get_expectation_suite(
              copy of _expectation_suite, not the original object.
         """
 
-        expectation_suite = copy.deepcopy(self._expectation_suite)
+        expectation_suite = copy.deepcopy(self.expectation_suite)
```
Comment (cdkini) [inline great_expectations/validator/validator.py:1421]: When working with a class internals, should we be using getters/setters? I think the private variable is totally fine here and is actually the better practice.

File: tests/rule_based_profiler/test_rule_based_profiler.py [hunk 4]
```diff
@@ -959,11 +959,23 @@ def test_run_profiler_with_dynamic_args(
...
+@mock.patch("great_expectations.rule_based_profiler.rule.Rule.generate")
+@mock.patch(
+    "great_expectations.rule_based_profiler.parameter_builder.ParameterBuilder.get_metrics"
+)
+@mock.patch(
+    "great_expectations.rule_based_profiler.parameter_builder.ParameterBuilder.get_batch_ids"
+)
+@mock.patch(
+    "great_expectations.rule_based_profiler.domain_builder.DomainBuilder.get_batch_ids"
+)
```
Comment (cdkini) [inline tests/rule_based_profiler/test_rule_based_profiler.py]: wow that's a **lot** of mocking - is there a better way to do this?

File: great_expectations/rule_based_profiler/rule_based_profiler.py [hunk 5]
```diff
@@ -360,6 +374,8 @@ def run(
                 expectation_suite._add_expectation(
                     expectation_configuration=expectation_configuration,
                     send_usage_event=False,
+                    match_type="domain",
```
Comment (cdkini) [inline great_expectations/rule_based_profiler/rule_based_profiler.py:375]: Enum?

File: great_expectations/rule_based_profiler/rule_based_profiler.py [hunk 6]
```diff
@@ -1044,40 +883,78 @@ def _generate_rule_overrides_from_batch_request(
...
-                domain_builder.batch_request = batch_request
-                domain_builder.batch_request["data_connector_query"] = {
-                    "index": -1,
-                }
+            if force_batch_data or domain_builder.batch_request is None:
+                setattr(
+                    domain_builder,
+                    batch_data_property_name,
+                    batch_data_property_value,
+                )
```
Comment (cdkini) [inline great_expectations/rule_based_profiler/rule_based_profiler.py]: noooooo can we avoid setattr at all?

File: great_expectations/rule_based_profiler/rule_based_profiler.py [hunk 7]
```diff
@@ -1232,6 +1109,189 @@ def list_profilers(
             return profiler_store.list_keys()
         return [x.configuration_key for x in profiler_store.list_keys()]
 
+    def self_check(self, pretty_print=True) -> dict:
```
Comment (cdkini) [inline great_expectations/rule_based_profiler/rule_based_profiler.py]: please annotate the boolean value

**Review by Shinnnyshinshin — CHANGES_REQUESTED** [review]
Thank you Alex this looks great. My request is that `BatchRequest` be allowed to be passed in a `Union[RuntimeBatchRequest, BatchRequest, dict]`, as we do in other parts of the codebase.  This would allow for the behavior to be consistent with other GE machinery.

File: great_expectations/validator/validator.py (hunk 2 above)
Comment (cdkini) [reply great_expectations/validator/validator.py]: Please raise Exception

File: tests/rule_based_profiler/test_rule_based_profiler.py (hunk 4 above)
Comment (cdkini) [reply tests/rule_based_profiler/test_rule_based_profiler.py]: try mocking ParameterBuilder?

File: great_expectations/data_context/data_context.py [hunk 8]
```diff
@@ -3410,7 +3410,8 @@ def run_profiler_with_dynamic_arguments(
     )
     def run_profiler_on_data(
         self,
-        batch_request: Union[dict, BatchRequest, RuntimeBatchRequest],
+        batch_list: Optional[List[Batch]] = None,
+        batch_request: Optional[Union[BatchRequest, RuntimeBatchRequest, dict]] = None,
```
Comment (Shinnnyshinshin) [inline great_expectations/data_context/data_context.py:3414]: ❤️

File: great_expectations/rule_based_profiler/expectation_configuration_builder/default_expectation_configuration_builder.py [hunk 9]
```diff
@@ -301,7 +301,7 @@ def _build_expectation_configuration(
         domain: Domain,
         variables: Optional[ParameterContainer] = None,
         parameters: Optional[Dict[str, ParameterContainer]] = None,
-    ) -> Union[ExpectationConfiguration, None]:
+    ) -> Optional[ExpectationConfiguration]:
```
Comment (Shinnnyshinshin) [inline great_expectations/rule_based_profiler/expectation_configuration_builder/default_expectation_configuration_builder.py:304]: Thank you for catching stuff like this @alexsherstinsky

File: great_expectations/rule_based_profiler/types/builder.py [hunk 10]
```diff
@@ -60,12 +66,40 @@ def batch_request(self) -> Optional[Union[BatchRequest, RuntimeBatchRequest, dic
...
+        if num_supplied_batch_specification_args > 1:
+            raise ge_exceptions.ProfilerConfigurationError(
+                f'Please pass at most one of "batch_list" and "batch_request" arguments (you passed {num_supplied_batch_specification_args} arguments).'
+            )
+
+        if batch_list is None:
+            self.batch_request = batch_request
+        else:
+            self.batch_list = batch_list
+
```
Comment (Shinnnyshinshin) [inline great_expectations/rule_based_profiler/types/builder.py:102]: In `builder.py`! Thank you for making this move @alexsherstinsky 

cc @cdkini

File: great_expectations/validator/validator.py (hunk 1 above)
Comment (Shinnnyshinshin) [inline great_expectations/validator/validator.py:576]: I really like this approach @alexsherstinsky thank you

File: great_expectations/validator/validator.py [hunk 11]
```diff
@@ -595,15 +584,19 @@ def _validate_rule_and_update_rule_properties(
...
-                exclude using active Batch (in order to avoid estimation bias).  However, when ParameterBuilder is part
-                of RuleBasedProfiler used to estimate arguments of an Expectation class, all Batch objects must be used.
-                """
-                parameter_builder.batch_list = list(self.batches.values())
 
+        domain_type: MetricDomainTypes = rule.domain_builder.domain_type
+        if domain_type not in MetricDomainTypes:
+            raise ValueError(
+                f'Domain type declaration "{domain_type}" in "MetricDomainTypes" does not exist.'
+            )
```
Comment (Shinnnyshinshin) [inline great_expectations/validator/validator.py:592]: ❤️

**Review by Shinnnyshinshin — APPROVED** [review]
LGTM @alexsherstinsky thank you very very much

**Pull request merged**